File: back/app/api/v1/drafts.py
"""Drafts and publication."""

from __future__ import annotations

from fastapi import APIRouter
from sqlalchemy import select

from app.api.deps import CurrentUser, DbSession
from app.api.schemas.common import Page
from app.api.schemas.works import (
    DraftCreateRequest,
    DraftResponse,
    LicenseInfo,
    PublishRequest,
    PublishResponse,
)
from app.domain.errors import Forbidden, NotFound, ValidationFailed
from app.domain.publishing import service as publishing
from app.models import Draft, LicenseSnapshot
from app.models.enums import Visibility
from app.presenters import media_urls

router = APIRouter(prefix="/drafts", tags=["drafts"])


@router.post("", response_model=DraftResponse, status_code=201)
def create_draft(
    payload: DraftCreateRequest, user: CurrentUser, session: DbSession
) -> DraftResponse:
    """Starts a draft, capturing the source licence at this moment."""
    draft = publishing.create_draft(
        session,
        user_id=user.id,
        source_work_id=payload.source_work_id,
        title=payload.title,
        params=payload.params,
    )
    session.commit()
    return _response(session, draft)


@router.get("", response_model=Page[DraftResponse])
def list_drafts(user: CurrentUser, session: DbSession) -> Page[DraftResponse]:
    drafts = session.scalars(
        select(Draft)
        .where(Draft.user_id == user.id, Draft.published_work_id.is_(None))
        .order_by(Draft.created_at.desc())
        .limit(50)
    )
    return Page(items=[_response(session, d) for d in drafts])


@router.get("/{draft_id}", response_model=DraftResponse)
def get_draft(draft_id: str, user: CurrentUser, session: DbSession) -> DraftResponse:
    return _response(session, _owned(session, draft_id, user.id))


@router.post("/{draft_id}/publish", response_model=PublishResponse, status_code=201)
def publish(
    draft_id: str, payload: PublishRequest, user: CurrentUser, session: DbSession
) -> PublishResponse:
    """Publishes a draft as one transaction.

    Everything from the licence re-check to the ancestor notification either
    lands together or not at all.
    """
    if not payload.ai_disclosure_confirmed:
        raise ValidationFailed(
            "请确认作品由 AI 生成的声明。", fields={"ai_disclosure_confirmed": "必须勾选"}
        )

    outcome = publishing.publish(
        session,
        user_id=user.id,
        draft_id=draft_id,
        title=payload.title,
        description=payload.description,
        visibility=payload.visibility,
        tags=payload.tags,
        cover_asset_id=payload.cover_asset_id,
        rights_confirmed=payload.rights_confirmed,
    )
    session.commit()

    return PublishResponse(
        work_id=outcome.work.id,
        work_version_id=outcome.version.id,
        visibility=Visibility(outcome.work.visibility),
        lineage_edge_id=outcome.lineage_edge.id if outcome.lineage_edge else None,
        royalties_paid=outcome.royalties,
    )


@router.delete("/{draft_id}", status_code=204)
def delete_draft(draft_id: str, user: CurrentUser, session: DbSession) -> None:
    draft = _owned(session, draft_id, user.id)
    if draft.published_work_id is not None:
        raise ValidationFailed("已发布的草稿不能删除。")
    session.delete(draft)
    session.commit()


def _owned(session, draft_id: str, user_id: str) -> Draft:  # type: ignore[no-untyped-def]
    draft = session.get(Draft, draft_id)
    if draft is None:
        raise NotFound("草稿不存在。")
    if draft.user_id != user_id:
        raise Forbidden("不能访问他人的草稿。")
    return draft


def _response(session, draft: Draft) -> DraftResponse:  # type: ignore[no-untyped-def]
    license_info = None
    if draft.license_snapshot_id:
        snapshot = session.get(LicenseSnapshot, draft.license_snapshot_id)
        if snapshot is not None:
            license_info = LicenseInfo(
                license_type=snapshot.license_type,
                attribution_text=snapshot.attribution_text,
                permissions=snapshot.permissions_json,
                captured_at=snapshot.captured_at,
            )

    return DraftResponse(
        id=draft.id,
        source_work_version_id=draft.source_work_version_id,
        title=draft.title,
        description=draft.description,
        params=draft.params_json,
        license=license_info,
        latest_job_id=draft.latest_job_id,
        output_asset_id=draft.output_asset_id,
        output_url=media_urls.asset_url(session, draft.output_asset_id),
        published_work_id=draft.published_work_id,
        created_at=draft.created_at,
    )

File: ios/Packages/ZaolangKit/Sources/ZaolangKit/Models/Drafts.swift
import Foundation

public struct DraftCreateRequest: Encodable, Sendable {
    public var sourceWorkID: String?
    public var title: String?
    public var params: [String: JSONValue]

    public init(sourceWorkID: String? = nil, title: String? = nil, params: [String: JSONValue] = [:]) {
        self.sourceWorkID = sourceWorkID
        self.title = title
        self.params = params
    }

    private enum CodingKeys: String, CodingKey {
        case sourceWorkID = "source_work_id"
        case title, params
    }
}

/// 工作台的落地状态：新建走 `sourceWorkVersionID == nil`，二创非 nil——与 `GenerationJobCreateRequest`
/// 里 `sourceWorkID` 的两态判断是同一条产品规则，只是这里存的是版本 id 不是作品 id。
public struct DraftResponse: Codable, Sendable, Equatable, Identifiable {
    public let id: String
    public let sourceWorkVersionID: String?
    public let title: String?
    public let description: String?
    public let params: [String: JSONValue]
    public let license: LicenseInfo?
    public let latestJobID: String?
    public let outputAssetID: String?
    public let outputURL: String?
    public let publishedWorkID: String?
    public let createdAt: Date

    private enum CodingKeys: String, CodingKey {
        case id
        case sourceWorkVersionID = "source_work_version_id"
        case title, description, params, license
        case latestJobID = "latest_job_id"
        case outputAssetID = "output_asset_id"
        case outputURL = "output_url"
        case publishedWorkID = "published_work_id"
        case createdAt = "created_at"
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = try c.decode(String.self, forKey: .id)
        sourceWorkVersionID = try c.decodeIfPresent(String.self, forKey: .sourceWorkVersionID)
        title = try c.decodeIfPresent(String.self, forKey: .title)
        description = try c.decodeIfPresent(String.self, forKey: .description)
        params = try c.decodeIfPresent([String: JSONValue].self, forKey: .params) ?? [:]
        license = try c.decodeIfPresent(LicenseInfo.self, forKey: .license)
        latestJobID = try c.decodeIfPresent(String.self, forKey: .latestJobID)
        outputAssetID = try c.decodeIfPresent(String.self, forKey: .outputAssetID)
        outputURL = try c.decodeIfPresent(String.self, forKey: .outputURL)
        publishedWorkID = try c.decodeIfPresent(String.self, forKey: .publishedWorkID)
        createdAt = try c.decode(Date.self, forKey: .createdAt)
    }

    public var isRemix: Bool { sourceWorkVersionID != nil }
}

/// `POST /v1/drafts/{id}/publish` 请求体。`visibility` 默认 `publicViewOnly`，
/// 两个确认框默认不勾——发布页的确认状态在 UI 层维护，提交前必须都为 true。
public struct PublishRequest: Encodable, Sendable {
    public var title: String
    public var description: String?
    public var visibility: Visibility
    public var tags: [String]
    public var coverAssetID: String?
    public var rightsConfirmed: Bool
    public var aiDisclosureConfirmed: Bool

    public init(
        title: String,
        description: String? = nil,
        visibility: Visibility = .publicViewOnly,
        tags: [String] = [],
        coverAssetID: String? = nil,
        rightsConfirmed: Bool = false,
        aiDisclosureConfirmed: Bool = false
    ) {
        self.title = title
        self.description = description
        self.visibility = visibility
        self.tags = tags
        self.coverAssetID = coverAssetID
        self.rightsConfirmed = rightsConfirmed
        self.aiDisclosureConfirmed = aiDisclosureConfirmed
    }

    private enum CodingKeys: String, CodingKey {
        case title, description, visibility, tags
        case coverAssetID = "cover_asset_id"
        case rightsConfirmed = "rights_confirmed"
        case aiDisclosureConfirmed = "ai_disclosure_confirmed"
    }
}

public struct PublishResponse: Codable, Sendable, Equatable {
    public let workID: String
    public let workVersionID: String
    public let visibility: RawOrUnknown<Visibility>
    public let lineageEdgeID: String?
    public let royaltiesPaid: [JSONValue]

    private enum CodingKeys: String, CodingKey {
        case workID = "work_id"
        case workVersionID = "work_version_id"
        case visibility
        case lineageEdgeID = "lineage_edge_id"
        case royaltiesPaid = "royalties_paid"
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        workID = try c.decode(String.self, forKey: .workID)
        workVersionID = try c.decode(String.self, forKey: .workVersionID)
        visibility = try c.decode(RawOrUnknown<Visibility>.self, forKey: .visibility)
        lineageEdgeID = try c.decodeIfPresent(String.self, forKey: .lineageEdgeID)
        royaltiesPaid = try c.decodeIfPresent([JSONValue].self, forKey: .royaltiesPaid) ?? []
    }
}

public struct VisibilityUpdateRequest: Encodable, Sendable {
    public var visibility: Visibility

    public init(visibility: Visibility) {
        self.visibility = visibility
    }
}
